zend_abstract_interface/methods/php5/methods.c

```diff
@@ -76,27 +74,77 @@ static bool z_call_method_without_args_ex(zval *object, zend_class_entry *ce, co
      */
     if (func->common.fn_flags & ZEND_ACC_ABSTRACT) {
         // "Cannot call abstract method %s::%s()"
-        return false;
+        goto error_exit;
+    }
+
+    zend_fcall_info_cache fcc = {0};
+    fcc.initialized = 1;
+    fcc.function_handler = func;
+    fcc.calling_scope = ce;
+    if (!object && (EG(called_scope) && instanceof_function(EG(called_scope), ce TSRMLS_CC))) {
+        fcc.called_scope = EG(called_scope);
+    } else {
+        fcc.called_scope = ce;
     }
 
-    bool ret = false;
+    zend_fcall_info fci = {0};
+    fci.size = sizeof(zend_fcall_info);
+    fci.no_separation = 1;
+    fci.retval_ptr_ptr = retval;
+
+    /* If 'object' is NULL, the caller wants to call this method statically. */
+    if (object) {
+        fci.object_ptr = object;
+        fcc.object_ptr = object;
+    }
+
+    /* Marking these as 'volatile' will ensure they survive a possible long jump
+     * from a zend_bailout without getting clobbered.
+     */
+    volatile bool should_bail = false;
+    volatile int call_fn_result = FAILURE;
 
     zai_sandbox sandbox;
     zai_sandbox_open(&sandbox);
 
-    zend_try {
-        /* We cannot use zend_call_method_with_0_params() here since it is a
-         * macro that masks sizeof() and therefore does not calculate the
-         * method len correctly without using a string literal.
-         */
-        zend_call_method(obj, ce, &func, method, method_len, retval, 0, NULL, NULL TSRMLS_CC);
-        /* An unhandled exception will not result in a zend_bailout if there is
-         * an active execution context. This is a failed call if an exception
-         * was thrown. The sandbox will clean up our mess when it closes.
+    /* We add the zval args directly from the variable arguments, va_arg(),
+     * instead of using zend_fcall_info_argv() because:
+     *
+     *   - We can NULL-check each arg.
+     *   - We circumvent an unnecessary reference count increase;
+     *     zend_call_function() already copies the zvals and increments the RC.
+     *   - We avoid an unnecessary heap allocation for the params.
+     *   - We avoid unnecessary calls to zend_fcall_info_args_clear().
+     */
+    if (argc > 0) {
+        zval *params[MAX_ARGS];
+        zval **param_ptrs[MAX_ARGS];
+        for (uint32_t i = 0; i < (uint32_t)argc; ++i) {
+            zval *arg = va_arg(argv, zval *);
+            if (!arg) {
+                zai_sandbox_close(&sandbox);
+                goto error_exit;
+            }
+            params[i] = arg;
+            param_ptrs[i] = &params[i];
+        }
+
+        fci.param_count = (uint32_t)argc;
+        fci.params = param_ptrs;
+
+        /* The zend_call_method() API only supports up to two arguments so we
+         * use zend_call_function() instead.
          */
-        ret = !EG(exception);
+        zend_try { call_fn_result = zend_call_function(&fci, &fcc TSRMLS_CC); }
+        zend_catch { should_bail = true; }
+        zend_end_try();
+    } else {
```

Comment:
redundant else branch, is exactly equal to the part at end end of the if, just move it out of the if.

Reply:
I know it looks weird, but in the back of my mind I thought this was intentional (looking at [ZAI functions](https://github.com/DataDog/dd-trace-php/pull/1300/files#diff-12dfce37c631fbed130a7c83f3c7318120d7aa98231eec5174f9368cadeaea95R62-R88)).

Sure enough, we ran into an [UBSan failure](https://github.com/DataDog/dd-trace-php/pull/1234#discussion_r634818384) that complained about a zero-length VLA back when we were using `params[argc]`. But now that we don't use a VLA, I'll move this out of the else block. We'll probably want to follow up with changing this on ZAI functions later as well.